### ISSUE TITLE
Unable to retrieve AWS role name

### ISSUE THREAD

**Issue by ahlag** [googleapis/google-cloud-python#15185]
Thanks for stopping by to let us know something could be better!

**PLEASE READ**: If you have a support contract with Google, please create an issue in the [support console](https://cloud.google.com/support/) instead of filing on GitHub. This will ensure a timely response.

Please run down the following list and make sure you've tried the usual "quick fixes":

  - Search the issues already opened: https://github.com/googleapis/google-auth-library-python/issues

If you are still having issues, please be sure to include as much information as possible:

#### Environment details

  - OS: AWS EC2
  - Python version: 3.9
  - pip version: 23.2.1
  - `google-auth` version: 2.22.0

#### Steps to reproduce

1. Run the following code after setting up the Workload Identity Federation by following [GCP Doc](https://cloud.google.com/iam/docs/workload-identity-federation-with-other-clouds#mappings-and-conditions) and [Youtube](https://www.youtube.com/watch?v=wgzuLEdHcO0). Replicated all the steps.
```
import google.auth
import os
from google.cloud import storage

#os.environ['GOOGLE_APPLICATION_CREDENTIALS'] = "./XX.json"
os.environ['GOOGLE_CLOUD_PROJECT'] = "XXXX"

client = storage.Client()
buckets = client.list_buckets()
for bucket in buckets:
    print(bucket.name)
```

Making sure to follow these steps will guarantee the quickest resolution possible.

Thanks!

**Comment by sassmith** [comment]
hey @ahlag ! I was following the same guides and running into the same error. I will share how I fixed this, hopefully it helps you too.

The google.auth library is trying to use the AWS Instance Metadata Service (IMDS) of your EC2 instance to grab the aws_role_name, aws_region, etc. 

There are 2 versions of IMDS, v1 is a request/response method and v2 is a session-oriented method, Google.auth supports both versions. To get the AWS region it is making a GET request to `http://169.254.169.254/latest/meta-data/placement/availability-zone` (the AWS metadata server) and will add the IMDSv2 session token to the headers of the request if present.

For me, the IMDSv2 session token was not being included in the request, making it a IMDSv1 request. However my EC2 instance was requiring IMDSv2, producing the same `401 Unauthorized` message you got. Changing my EC2 instance to IMDSv2 optional fixed this issue.

Steps to fix:
- Go to your EC2 instance in AWS
- Actions -> Instance Settings -> Modify instance metadata options

<img width="420" alt="Screenshot 2023-08-22 at 4 43 37 PM" src="https://github.com/googleapis/google-cloud-python/assets/97194436/755fee14-5ba0-4c13-8a63-56e1827e8d76">

<img width="489" alt="Screenshot 2023-08-22 at 3 31 54 PM" src="https://github.com/googleapis/google-cloud-python/assets/97194436/1189cc2e-504b-41f3-b784-e6617a8558b5">

If your organization requires IMDSv2 be required then I unfortunately don't know the steps needed to get around that but hopefully this gets you a little bit closer!

Long story short: I think this is an issue with config on the AWS side which most of these Workload Identity Federation guides brush over and not actually a bug with the Google.auth library.

**Comment by PoorvaRaut** [comment]
I am Poorva Raut and I am interested in contributing to your project regarding this issue. So can I contribute? If yes, kindly guide me on how to get started or share contribution guidelines

**Comment by Wayne9981** [comment]
According to the [gcloud documentation](https://cloud.google.com/sdk/gcloud/reference/iam/workload-identity-pools/create-cred-config#--enable-imdsv2), Workload Identity Federation AWS provider now seems to support IMDSv2.

**Comment by parthea** [comment]
Transferring to google-cloud-python as the code for google-auth-library-python will be migrated there in 1-3 weeks.

**Comment by daniel-sanche** [comment]
This issue is being bulk closed due to inactivity. Please re-open if you think this still requires attention, and we'll take another look. Thanks!
